Add grpc samples

qu1queee opened the pull request:
Introduce grpc directory, containing a microservice architecture consisting of an internal grpc server, and a public client/server app.

A run.sh script is provided, to easily deploy this architecture in Code Engine, as well as a README, explaining the ideas behind.

reggeenr commented on grpc/run.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
Can you please rename the script to "just" `run`? This way it will nicely co-exists next to the other existing run scripts. See: https://github.com/IBM/CodeEngine/blob/main/bin/run-all

reggeenr commented on grpc/run.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+set -euo pipefail
```
In order to allow it being executed as part of test automation and to prevent users from having components being installed that incur costs, without the users understanding the full impact of them running the script, we'll need to add a cleanup routine to this script. See https://github.com/IBM/CodeEngine/blob/main/function2job/run for an example.

reggeenr commented on grpc/client/main.go:
```diff
@@ -0,0 +1,121 @@
+
+func main() {
+
+	serverLocalEndpoint := "LOCAL_ENDPOINT_WITH_PORT"
+
+	localEndpoint := os.Getenv(serverLocalEndpoint)
```
```suggestion
	localEndpoint := os.Getenv("LOCAL_ENDPOINT_WITH_PORT")
```

reggeenr commented on grpc/client/main.go:
```diff
@@ -0,0 +1,121 @@
+	fmt.Printf("using local endpoint: %s\n", localEndpoint)
+	conn, err := grpc.Dial(localEndpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		log.Fatalf("failed to connect: %v", err)
```
Shouldn't we and stop the app initialization at this point?

qu1queee replied:
log.Fatalf comes with an os.Exit(1)

reggeenr commented on grpc/Dockerfile.client:
```diff
@@ -0,0 +1,21 @@
+FROM golang:latest AS stage
```
To avoid image pull throttling, we are serving all images that our samples require from our public ICR namespace. 
```suggestion
FROM icr.io/codeengine/golang:latest AS stage
```

reggeenr commented on grpc/Dockerfile.client:
```diff
@@ -0,0 +1,21 @@
+
+RUN CGO_ENABLED=0 GOOS=linux go build -o client ./client
+
+FROM golang:latest
```
```suggestion
FROM icr.io/codeengine/golang:latest
```

reggeenr commented on grpc/Dockerfile.server:
```diff
@@ -0,0 +1,21 @@
+FROM golang:latest AS stage
```
```suggestion
FROM icr.io/codeengine/golang:latest AS stage
```

reggeenr commented on grpc/Dockerfile.server:
```diff
@@ -0,0 +1,21 @@
+
+RUN CGO_ENABLED=0 GOOS=linux go build -o server ./server
+
+FROM golang:latest
```
```suggestion
FROM icr.io/codeengine/golang:latest
```

reggeenr commented on grpc/README.md:
```diff
@@ -0,0 +1,39 @@
+Check the source code if you want to understand how this works. In general,
+we provided three directories:
+
+- `/ecommerce` directory hosts the protobuf files and declares the `grocery`
```
Is there every a need to regenerate this directory (e.g. after user made some adjustments to play around)? If so, I suggest to add the command to re-generate the files.

qu1queee replied:
done. Added a new script call `./regenerate-grpc-code.sh

reggeenr commented on grpc/README.md:
````diff
@@ -0,0 +1,39 @@
+Once you have selected your Code Engine project, you only need to run:
+
+```sh
+./run.sh
````
See comment below in the `run.sh` file

```suggestion
./run
```

reggeenr commented on grpc/run.sh:
```diff
@@ -0,0 +1,31 @@
+
+# Buy an item from electronics and pay with 2000
+echo "[INFO] Going to buy an iphone"
+curl -q "${URL}"/buygrocery/electronics/iphone/2000.0
```
I suggest to add a command that checks whether the buy operation actually succeeded. If it doesn't we should end the script with exit code non-zero

qu1queee replied:
covered

reggeenr reviewed (changes requested):
Hi @qu1queee , thank you for putting together this sample! Looking forward to get it in finally. I do have a few comments mainly related to the run script, which I would like to discuss with you. 

Next, I suggest that we plan a pairing session.

Best regards, Rico

qu1queee commented:
@reggeenr thanks for the review. I addressed all your comments. In addition, I added the following:

- A `build` script for pushing the container images of the client and server
- An architecture diagram in the README for this sample app

reggeenr reviewed (approved):
LGTM - Great job, @qu1queee ! Appreciate the work you put in to finish this sample.
